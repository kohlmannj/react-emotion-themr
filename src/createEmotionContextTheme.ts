import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import {
  createEmotionTheme,
  TReactEmotionThemrTheme,
  RenderedEmotionTheme,
} from './createEmotionTheme';

export type TReactEmotionThemrContextTheme = {
  [key: string]: TReactEmotionThemrTheme;
};

export function createEmotionContextTheme<T>(
  contextThemeObj: T
): RenderedEmotionTheme<T, TReactEmotionThemrContextTheme> {
  if (
    typeof contextThemeObj === 'object' &&
    typeof contextThemeObj !== 'undefined' &&
    contextThemeObj !== null
  ) {
    return Object.keys(contextThemeObj).reduce(
      (emotionContextTheme, componentName) => ({
        ...emotionContextTheme,
        [componentName]: createEmotionTheme(
          (contextThemeObj as TReactEmotionThemrTheme)[componentName]
        ),
      }),
      {} as TReactCSSThemrTheme
    ) as T extends TReactEmotionThemrContextTheme ? TReactCSSThemrTheme : never;
  }

  return contextThemeObj as T extends TReactEmotionThemrContextTheme ? never : T;
}
