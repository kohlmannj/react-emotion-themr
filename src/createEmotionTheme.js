import { css } from 'emotion';

export function createEmotionTheme(classNamesObj) {
  if (typeof classNamesObj !== 'object' || classNamesObj === null) {
    return classNamesObj;
  }

  return Object.keys(classNamesObj).reduce((emotionTheme, className) => {
    const value = classNamesObj[className];
    const typeofValueIsNonNullObject = typeof value === 'object' && value !== null;

    // Only add the className if `value` is a non-null object or a string
    if (typeofValueIsNonNullObject || typeof value === 'string') {
      return {
        ...emotionTheme,
        [className]: typeofValueIsNonNullObject ? css(value, `label:${className};`) : value,
      };
    }

    return emotionTheme;
  }, {});
}
