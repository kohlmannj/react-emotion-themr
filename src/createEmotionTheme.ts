import { TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import { CSSObject } from 'create-emotion';
import { css } from 'emotion';

export type TReactEmotionThemrTheme = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: CSSObject | any;
};

export type RenderedEmotionTheme<T, V = TReactEmotionThemrTheme> = T extends V
  ? TReactCSSThemrTheme
  : T;

export function createEmotionTheme<T>(classNamesObj: T): RenderedEmotionTheme<T> {
  if (
    typeof classNamesObj === 'object' &&
    typeof classNamesObj !== 'undefined' &&
    classNamesObj !== null
  ) {
    return Object.keys(classNamesObj).reduce(
      (emotionTheme, className) => {
        const value = (classNamesObj as TReactEmotionThemrTheme)[className];

        if ((typeof value === 'object' && value !== null) || typeof value === 'string') {
          return {
            ...emotionTheme,
            [className]:
              typeof value === 'object' && value !== null
                ? css(value, `label:${className};`)
                : value,
          } as TReactCSSThemrTheme;
        }

        return emotionTheme;
      },
      {} as TReactCSSThemrTheme
      /** @see https://github.com/microsoft/TypeScript/issues/22735#issuecomment-376960435 */
    ) as T extends TReactEmotionThemrTheme ? TReactCSSThemrTheme : never;
  }

  return classNamesObj as T extends TReactEmotionThemrTheme ? never : T;
}
