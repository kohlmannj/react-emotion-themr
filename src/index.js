export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themr';
