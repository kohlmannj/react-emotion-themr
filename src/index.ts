export * from './createEmotionContextTheme';
export * from './createEmotionTheme';
export * from './ThemeProvider';
export * from './themeable';
export * from './themr';
