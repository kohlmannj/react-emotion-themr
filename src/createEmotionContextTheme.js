import { createEmotionTheme } from './createEmotionTheme';

export function createEmotionContextTheme(contextThemeObj) {
  if (typeof contextThemeObj !== 'object' || contextThemeObj === null) {
    return contextThemeObj;
  }

  return Object.keys(contextThemeObj).reduce(
    (emotionContextTheme, componentName) => ({
      ...emotionContextTheme,
      [componentName]: createEmotionTheme(contextThemeObj[componentName]),
    }),
    {}
  );
}
