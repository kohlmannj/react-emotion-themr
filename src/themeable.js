import { themeable as cssThemeable } from '@friendsofreactjs/react-css-themr';
import { createEmotionTheme } from './createEmotionTheme';

/**
 * Merges passed themes by concatenating string keys and processing nested themes
 *
 * @param {...TReactCSSThemrTheme} themes - Themes
 * @returns {TReactCSSThemrTheme} - Resulting theme
 */
export function themeable(...themes) {
  return cssThemeable(...themes.map(createEmotionTheme));
}
