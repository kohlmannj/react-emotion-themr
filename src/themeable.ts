import { themeable as cssThemeable, TReactCSSThemrTheme } from '@friendsofreactjs/react-css-themr';
import { createEmotionTheme, TReactEmotionThemrTheme } from './createEmotionTheme';

/**
 * Merges passed themes by concatenating string keys and processing nested themes
 *
 * @param themes - Themes
 * @returns {TReactCSSThemrTheme} - Resulting theme
 */
export function themeable(...themeObjects: TReactEmotionThemrTheme[]): TReactCSSThemrTheme {
  const themes = [];
  for (let i = 0; i < themeObjects.length; i++) {
    themes.push(createEmotionTheme(themeObjects[i]));
  }
  return cssThemeable.call(null, ...themes);
}
