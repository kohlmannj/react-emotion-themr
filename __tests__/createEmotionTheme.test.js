import * as emotion from 'emotion';
import { getStyles } from 'jest-emotion';
import { createEmotionTheme } from '../src/createEmotionTheme';

function objectContainingClassNames(themeObj) {
  return expect.objectContaining(
    Object.keys(themeObj).reduce(
      (expectedObj, key) => ({
        ...expectedObj,
        ...(typeof themeObj[key] === 'string' ||
        (typeof themeObj[key] === 'object' && themeObj[key] !== null)
          ? { [key]: typeof themeObj[key] === 'string' ? themeObj[key] : expect.any(String) }
          : {}),
      }),
      {}
    )
  );
}

afterEach(() => {
  emotion.flush();
});

describe('createEmotionTheme', () => {
  it('renders a theme of objects', () => {
    const objectOfObjectsTheme = {
      root: {
        backgroundColor: 'rebeccapurple',
      },
      foo: {
        fontSize: 32,
      },
    };

    const emotionTheme = createEmotionTheme(objectOfObjectsTheme);

    expect(emotionTheme).toEqual(objectContainingClassNames(objectOfObjectsTheme));
    expect(getStyles(emotion)).toMatchSnapshot();
  });

  it(`handles mixed themes (an object of strings and non-null objects)`, () => {
    const mixedTheme = {
      root: {
        backgroundColor: 'rebeccapurple',
      },
      foo: 'foo',
    };

    const emotionTheme = createEmotionTheme('foobar');

    expect(emotionTheme).toBe('foobar');
    const styles = getStyles(emotion);
    expect(styles).not.toContain(mixedTheme.root.backgroundColor);
  });

  it(`omits properties whose values are neither strings nor non-null objects`, () => {
    const themeWithNullProperties = {
      root: 'root',
      foo: null,
    };

    const emotionTheme = createEmotionTheme(themeWithNullProperties);

    expect(emotionTheme).not.toHaveProperty('foo');
    const styles = getStyles(emotion);
    expect(styles).toBe('');
  });

  it(`passes through a theme of strings (and doesn't render anything with Emotion)`, () => {
    const objectOfStringsTheme = {
      root: 'root',
      foo: 'foo',
    };

    const emotionTheme = createEmotionTheme(objectOfStringsTheme);

    expect(emotionTheme).toEqual(objectOfStringsTheme);
    const styles = getStyles(emotion);
    expect(styles).toBe('');
  });

  it(`returns the input if it's a non-object`, () => {
    const emotionTheme = createEmotionTheme('foobar');

    expect(emotionTheme).toBe('foobar');
    const styles = getStyles(emotion);
    expect(styles).toBe('');
  });
});
