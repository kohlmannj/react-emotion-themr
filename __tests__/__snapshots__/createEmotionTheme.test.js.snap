// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEmotionTheme renders a theme of objects 1`] = `".css-ux9s12-root{background-color:rebeccapurple;}.css-1pzc2nm-foo{font-size:32px;}"`;
